File: downloader.py
import logging
from typing import Iterator, List
from http_client import HttpClient
from ogg_parser import reconstruct_ogg_from_chunks
from clients.playplay_client import PlayPlayClient
from playplay_emulator.playplay_keygen import PlayPlayKeygen
from proto.track_pb2 import AudioFile, Track
from clients.storage_resolve_client import StorageResolverClient

CHUNK_SIZE = 0x10000
logger = logging.getLogger(__name__)

AUDIO_FORMATS = {
    "ogg-160": AudioFile.Format.OGG_VORBIS_160,
    "ogg-96": AudioFile.Format.OGG_VORBIS_96,
    "ogg-320": AudioFile.Format.OGG_VORBIS_320,
}


def download_decrypt_and_reconstruct(
    http: HttpClient,
    url: str,
    keygen: PlayPlayKeygen,
) -> Iterator[bytes]:
    with http.stream(url, headers={"Range": "bytes=0-"}) as resp:
        resp.raise_for_status()

        decrypted_chunks = (
            buf
            for buf in keygen.decrypt_stream(
                bytearray(chunk)
                for chunk in resp.iter_content(chunk_size=CHUNK_SIZE)
                if chunk
            )
            if buf
        )

        yield from reconstruct_ogg_from_chunks(decrypted_chunks)


def download_track(
    http_client: HttpClient,
    track: Track,
    resolver: StorageResolverClient,
    playplay: PlayPlayClient,
    keygen: PlayPlayKeygen,
    audio_format: AudioFile.Format,
) -> None:
    file_id = next(
        (audio.file_id for audio in track.file if audio.format == audio_format),
        None,
    )

    if not file_id:
        logger.warning("Audio format unavailable for this track")
        return

    obfuscated_key = playplay.get_obfuscated_key(file_id)
    keygen.configure(file_id=file_id, obfuscated_key=obfuscated_key)

    urls = resolver.resolve(file_id)
    if not urls:
        raise RuntimeError("No URL returned by the resolver")

    output_path = f"{file_id.hex()}.ogg"
    logger.info("Downloading: %s", output_path)

    with open(output_path, "wb") as f:
        for ogg_page in download_decrypt_and_reconstruct(
            http_client,
            urls[0],
            keygen,
        ):
            f.write(ogg_page)

File: src/downloader.rs
use anyhow::Context;
use tracing::info;

use crate::clients::{playplay::PlayPlayClient, storage_resolve::StorageResolverClient};
use crate::http_client::HttpClient;
use crate::ogg_parser::reconstruct_ogg_from_chunks;
use crate::playplay_keygen::PlayPlayKeygen;
use crate::proto::track::audio_file::Format as AudioFormat;
use crate::proto::track::Track;

use clap::ValueEnum;

#[derive(Copy, Clone, Debug, Eq, PartialEq, ValueEnum)]
pub enum Quality {
    #[value(name = "ogg-96")]
    Ogg96,
    #[value(name = "ogg-160")]
    Ogg160,
    #[value(name = "ogg-320")]
    Ogg320,
}

#[derive(Clone, Copy)]
pub struct AudioFormatChoice {
    pub format: AudioFormat,
}

pub fn audio_format(quality: Quality) -> AudioFormatChoice {
    match quality {
        Quality::Ogg96 => AudioFormatChoice {
            format: AudioFormat::OggVorbis96,
        },
        Quality::Ogg160 => AudioFormatChoice {
            format: AudioFormat::OggVorbis160,
        },
        Quality::Ogg320 => AudioFormatChoice {
            format: AudioFormat::OggVorbis320,
        },
    }
}

const CHUNK_SIZE: usize = 0x10000;

pub fn download_track(
    http: &HttpClient,
    track: &Track,
    resolver: &StorageResolverClient,
    playplay: &PlayPlayClient,
    keygen: &PlayPlayKeygen,
    audio_format: AudioFormatChoice,
) -> anyhow::Result<()> {
    let file_id = track
        .file
        .iter()
        .find(|f| f.format == Some(audio_format.format as i32))
        .and_then(|f| f.file_id.clone())
        .context("audio format unavailable")?;

    let obfuscated_key = playplay.get_obfuscated_key(&file_id)?;
    let mut kg = keygen.clone();
    kg.configure(&file_id[..16], &obfuscated_key)?;

    let urls = resolver.resolve(&file_id)?;
    let url = urls
        .get(0)
        .cloned()
        .context("No URL returned by resolver")?;

    let output_path = format!("{}.ogg", hex::encode(&file_id));
    info!("Downloading: {}", output_path);

    let mut file = std::fs::File::create(&output_path)?;
    let mut resp = http.stream(&url)?;
    let body = resp.bytes()?;
    let decrypted = kg.decrypt_stream(std::iter::once(body.to_vec()));
    for page in reconstruct_ogg_from_chunks(decrypted) {
        std::io::Write::write_all(&mut file, &page)?;
    }

    Ok(())
}
